doc/api/async_hooks.md

```diff
@@ -950,6 +950,33 @@ If this method is called outside of an asynchronous context initialized by
 calling `asyncLocalStorage.run` or `asyncLocalStorage.runAndReturn`, it will
 return `undefined`.
 
+### `asyncLocalStorage.enterWith(store)`
+<!-- YAML
+added: REPLACEME
+-->
+
+* `store` {any}
+
+Calling `asyncLocalStorage.enterWith(store)` will transition into the context
+for the remainder of the current synchronous execution and will persist
+through any following asynchronous calls.
+
+Example:
+
+```js
+const store = { id: 1 };
+asyncLocalStorage.enterWith(store);
+asyncLocalStorage.getStore(); // Returns the store object
+someAsyncOperation(() => {
+  asyncLocalStorage.getStore(); // Returns the same object
+});
+```
+
+This transition will continue for the _entire_ synchronous execution.
```

Comment:
This is actually not true if `enterWith()` is called within `run()` as there is no stack, just a value.

Reply:
I don't _intend_ for it to be used within a `run(...)`. It's an _either/or_ thing. I don't think it makes sense to use one inside the other. :thinking: 

Not sure I understand what you mean about "no stack, just a value" though.

Anyway, my _intent_ is for this to be an _alternative_ to the `run(...)` methods for when you want to ensure the context _transitions_ for the remaining lifetime of the sync code and all subsequent async branches. It's intended for APMs that want to be able to encapsulate code they don't control and therefore would otherwise have to apply some gross and potentially brittle monkey-patches to achieve the same thing.

Reply:
I understand that it is the intention to use either `run` or `enterWith()` and it's not built for being use inside the callback passed to `run()`.

But it's not ensured nor does the user get any hint about wrong usage. I have usecases in mind where a high layer abstraction of some driver does sync calls to an lower layer. Both are instrumented by an APM as the lower layer can be use also standalone.
I think in such cases it may happen easily that `enterWith()` is called twice or inside a `run()` callback.

The question is should we disallow this? Warn the user about it? or just document what happens. A user can always do a `getStore()` first to check if a context is active or not.

Regarding stack/value: Currently a nested `run` or `enterWith()` overwrites the store in the resource. So after the inner `run` is done the store is undefined. Using as stack would restore the previous store.

Reply:
I would be globally -1 for having a stack here for now :/ It has been a great source of complexity with domains.

I'd say that I would rather extend documentation for that case and posuh toward the following pattern is that case:
```js
enterWith(getStore() || my_new_store)
```

wdyt?

Reply:
Why not just have it throw if you try to use enter/run when already entered/running? Seems like universally _wrong_ behaviour to me. :thinking:

Reply:
This is also an issue even _without_ exposing `enterWith(store)` a run within another run will blow away the old value.

```js
cls.run(() => {
  cls.getStore().set('foo', 'bar')
  cls.run(() => {
    cls.getStore().get('foo') // error
  })
})
```

Reply:
I would like this API to throw at least as possible altogether:

```js
try {
    enterWith({})
} catch(e) {
}
// here, I need to do some deep dive to know what the state here
```
I much prefer the overwrite behavior as it has the least surprise (to me at least).

Reply:
If overwriting is the _preferred_ behaviour, then the action with `enterWith(store)` is "do nothing" as its just a state transition from one store to another. :man_shrugging:

Reply:
Yes, nesting is already problematic now, even this fails:
```js
cls.runSyncAndReturn(() => {
  cls.getStore().set('foo', 'bar')
  cls.runSyncAndReturn(() => {})
  cls.getStore().get('foo') // it's not bar
})
```
The `enterWith(getStore() || my_new_store)` pattern won't solve above problem.

I think throwing is better then silently overwriting/clearing. I can imaging that it may be incredible hard to debug why a store is missing at some place. Could be normal context loss or silent overwrite...

An overwriting `enterWith()` sounds somehow ok as it is a transition. Users could implement a stack on their own if really needed. Seems the statement below that the actual problem is `exit` holds true.

By the way, `enterWith(undefined)` is actually an exit. Is this intended?

But in any case it should be documented.

Reply:
@Flarna what about
```js
cls.runSyncAndReturn(new Map(), () => {
  cls.getStore().set('foo', 'bar')
  cls.runSyncAndReturn(cls.getStore() || new Map(), () => {})
  cls.getStore().get('foo') // it's not bar
})
```

this is what we will have once https://github.com/nodejs/node/pull/31930 lands.

> By the way, enterWith(undefined) is actually an exit. Is this intended?

I did not catch that. That's indeed problematic but probably falls on https://github.com/nodejs/node/pull/31930

Reply:
@Flarna If your original comment was referring to the `this._exit()` call from an outer `run(...)` blowing away the value, this comment is relevant, I think: https://github.com/nodejs/node/pull/31945#issuecomment-590800282

The state transition moves, from the barrier between sync callbacks to the exit trigger, buts its functionally the same as if it there was only an enter transition in a nested async call. As stated in the referenced comment, the sync version even could just be expressed with `AsyncResource` and again would not need the exit trigger. I think the behaviour here is _reasonable_, but perhaps something could be said about how the `run*(...)` methods will _create_ this encapsulation. :thinking:

Reply:
> By the way, enterWith(undefined) is actually an exit. Is this intended?

It's true, but I don't see a big impact as propagation is ruled by the `enabled` property. But thinking about it, this is probably a mistake here, we would want a way not to propagate `undefined` indefinitly other than using `disable()`.

Reply:
Exit could be changed to something like this:

```js
  exit(callback, ...args) {
    const store = this.getStore();
    this.disable();
    process.nextTick(callback, ...args);
    this.enterWith(store);
  }
```

That'd eliminate the need for the `enabled` tracking property.

Reply:
Sorry, but I think my last statements were quite unclear as I often wrote just `run()` but actually meant `runSyncAndReturn()`.

Nesting the async `run` is not problem as actually they are not really nested. Mixing it with `enterWith()` should be quite ok as the `enterWith()` just move from one store to the next.

Problematic seems to be only the nesting of `runSyncAndReturn()`, independent of `enterWith()`. I haven't discovered this during the initial PR, just noticed it here even it's not caused by this PR. Maybe better to move this discussion to a separated issue.

Reply:
@Flarna feel free to open a PR so we can iterate :)

Reply:
@vdeturckheim I will wait on the result of this one first.

Reply:
> Problematic seems to be only the nesting of `runSyncAndReturn()`, independent of `enterWith()`.

Definitely sounds like something that should be improved. I also forgot to take a look at that behavior in the original PR.

Reply:
I’m on it. Looking at a few other things anyway.

Reply:
I've opened #31950 to deal with that. I probably could've just done it here, but I don't want to mix too much potentially controversial changes that could slow down getting them merged.

Reply:
Can I mark this as resolved, or is there still some action here? If #31950, the concern raised is no longer correct as the exits are eliminated, so I believe my current statement in the docs is correct.

Reply:
Feel free to resolve